include/envoy/ssl/context_manager.h

```diff
@@ -38,6 +39,11 @@ class ContextManager {
    * Iterate through all currently allocated contexts.
    */
   virtual void iterateContexts(std::function<void(const Context&)> callback) PURE;
+
+  /**
+   * Return the instance of secret manager.
+   */
+  virtual Secret::SecretManager& secretManager() PURE;
```

Comment:
Actually, looking at this again, what's the reasoning for `Ssl::ContextManager` having `SecretManager()`?
Should you use the global one from `InstanceImpl->secretManager()`?

Reply:
Consumer of the secret manager is the createTransportSocketFactory of UpstreamSslSocketFactory and DownstreamSslSocketFactory.

sslContextManager()  of Server::Configuration::TransportSocketFactoryContext& context is 
available in that function. 

Is there a direct access to the global one?

Reply:
Looking at the code, I'd rather move it one level up and put it in `Configuration::TransportSocketFactoryContext` next to the `sslContextManager()`. That way, it's still going to be accessible in `createTransportSocketFactory()`.

Note: perhaps once you add dynamic SDS there will be a justification for tying `Ssl::ContextManager` and `Secret::SecretManager` together (although, it would be probably the other way around), but I don't think there is a reason to do that just yet.

Reply:
ListenerImpl and ClusterInfoImpl are implementations of Configuration::TransportSocketFactoryContext interface. 
If I need to add secretManager() to those class, I need to update constructors of derived classes and instance creation codes .
Adding secretManager() to Ssl::ContextManagerImpl seems to be easy to use.
Please let me know your opinion.

Reply:
> ListenerImpl and ClusterInfoImpl are implementations of Configuration::TransportSocketFactoryContext interface.
If I need to add secretManager() to those class, I need to update constructors of derived classes and instance creation codes .

Adding `secretManager()` to listeners should be as easy as adding this line:
```
Secret::SecretManager& secretManager() override { return parent_.server_.secretManager(); }
```

For clusters, you'd indeed need to change the constructor, and I'd even say that you might change it to pass `ClusterManagerImpl`, to match listener, and then you should be able to return `parent_.server_.secretManager()` as well (perhaps in separate PR).

cc @mattklein123 for thoughts on this.

Reply:
I'm fine either way. If @PiotrSikora has thought about it and thinks it's better I would just take that approach.

Reply:
ok I am moving the method to the Configuration::TransportSocketFactoryContext interface. 
It requires 25 files changes and related test cases. It would take some time.